libs/utils/src/properties_encoding.c

```diff
@@ -0,0 +1,580 @@
+        break;
+    case CELIX_PROPERTIES_VALUE_TYPE_BOOL:
+        *out = json_boolean(entry->typed.boolValue);
+        break;
+    case CELIX_PROPERTIES_VALUE_TYPE_VERSION:
+        return celix_properties_versionToJson(entry->typed.versionValue, out);
+    case CELIX_PROPERTIES_VALUE_TYPE_ARRAY_LIST:
+        return celix_properties_arrayEntryValueToJson(key, entry, flags, out);
+    default:
+        // LCOV_EXCL_START
+        celix_err_pushf("Unexpected properties entry type %d.", entry->valueType);
+        return CELIX_ILLEGAL_ARGUMENT;
+        // LCOV_EXCL_STOP
+    }
+
+    if (!*out) {
+        celix_err_pushf("Failed to create json value for key '%s'.", key);
+        return ENOMEM;
+    }
+    return CELIX_SUCCESS;
+}
+
+static celix_status_t celix_properties_addJsonValueToJson(json_t* value, const char* key, json_t* obj, int flags) {
+    if (!value) {
+        // ignore unset values
+        return CELIX_SUCCESS;
+    }
+
+    json_t* field = json_object_get(obj, key);
+    if (field) {
+        if (flags & CELIX_PROPERTIES_ENCODE_ERROR_ON_COLLISIONS) {
+            celix_err_pushf("Invalid key collision. key '%s' already exists.", key);
+            return CELIX_ILLEGAL_ARGUMENT;
+        }
+    }
+
+    int rc = json_object_set_new(obj, key, value);
+    if (rc != 0) {
+        celix_err_push("Failed to set json object");
+        return ENOMEM;
+    }
+    return CELIX_SUCCESS;
+}
+
+static celix_status_t celix_properties_addPropertiesEntryFlatToJson(const celix_properties_entry_t* entry,
+                                                                    const char* key,
+                                                                    json_t* root,
+                                                                    int flags) {
+    json_t* value;
+    celix_status_t status = celix_properties_entryValueToJson(key, entry, flags, &value);
+    status = CELIX_DO_IF(status, celix_properties_addJsonValueToJson(value, key, root, flags));
+    return status;
+}
+
+static celix_status_t celix_properties_addPropertiesEntryToJson(const celix_properties_entry_t* entry,
+                                                                const char* key,
+                                                                json_t* root,
+                                                                int flags) {
+    json_t* jsonObj = root;
+    const char* fieldName = key;
+    const char* slash = strstr(key, "/");
+    while (slash) {
+        char buf[64];
+        char* name = celix_utils_writeOrCreateString(buf, sizeof(buf), "%.*s", (int)(slash - fieldName), fieldName);
+        celix_auto(celix_utils_string_guard_t) strGuard = celix_utils_stringGuard_init(buf, name);
+        if (!name) {
+            celix_err_push("Failed to create name string");
+            return ENOMEM;
+        }
+        json_t* subObj = json_object_get(jsonObj, name);
+        if (!subObj || !json_is_object(subObj)) {
```

Comment:
It fails the following test:

```C++
celix_autoptr(celix_properties_t) props = celix_properties_create();
celix_properties_set(props, "key1", "value1");
celix_properties_set(props, "key2", "value2");
celix_properties_set(props, "object1/key3", "value3");
celix_properties_set(props, "object1/key4", "value4");
celix_properties_set(props, "object2/key5", "value5");
celix_properties_set(props, "object3/object4/key6", "value6");

//And an in-memory stream
celix_autofree char* output;

//When saving the properties to the stream
auto status = celix_properties_saveToString(props, CELIX_PROPERTIES_ENCODE_NESTED_STYLE | CELIX_PROPERTIES_ENCODE_ERROR_ON_COLLISIONS, &output);
ASSERT_EQ(CELIX_SUCCESS, status);
```

Reply:
If `subObj` is not an object, it makes no sense to try to insert key-value pair into it, which merely leads to errors.
The fix still fails the following test

```C++
    celix_autoptr(celix_properties_t) props2 = celix_properties_create();
    // pick keys such that key1 appears before key2 when iterating over the properties
    celix_properties_set(props2, "a.b.c", "value1");
    celix_properties_set(props2, "a.b.c.d", "value2");
    CELIX_PROPERTIES_ITERATE(props2, iter) {
        std::cout << "ZPTest " << iter.key << std::endl;
    }
    status = celix_properties_saveToString(props2, CELIX_PROPERTIES_ENCODE_NESTED_STYLE, &output);
    EXPECT_EQ(CELIX_SUCCESS, status);
```

Note that the above is included in `TEST_F(PropertiesSerializationTestSuite, KeyCollision)`.
Please check my fix.